File: 백준/Gold/16928. 뱀과 사다리 게임/뱀과 사다리 게임.cc
#include <bits/stdc++.h>
using namespace std;
int n, m;            // 사다리 수, 뱀의 수
int x, y;            // 사다리의 정보
int u, v;            // 뱀의 정보
int dist[101];       
int ladder[101];     // 사다리
int bam[101];        // 뱀
bool vis[101];       
int main() {
    cin >> n >> m;
    // 사다리 정보
    for(int i=0; i<n; i++) {
        cin >> x >> y; // x번칸 도착 y번칸 이동
        ladder[x] = y;
    }
    // 뱀의 정보
    for(int i=0; i<m; i++) {
        cin >> u >> v; // u번칸 도착하면 v번칸 이동
        bam[u] = v;
    }
    queue<int> q;
    q.push(1);
    vis[1] = true;
    dist[1] = 0;
    while(!q.empty()) {
        auto cur = q.front(); q.pop();
        if(cur == 100) {
            cout << dist[cur];
            return 0;
        }
        for(int dice=1; dice<=6; dice++) {
            int nx = cur + dice; 
            if(nx > 100) continue;
            if(ladder[nx]) nx = ladder[nx];
            if(bam[nx]) nx = bam[nx];
            if(!vis[nx]) {
                vis[nx] = true;
                dist[nx] = dist[cur] + 1;
                q.push(nx);
            }
        }
    }
}

File: 백준/Gold/16928. 뱀과 사다리 게임/뱀과 사다리 게임.java
import java.util.*;
import java.lang.*;
import java.io.*;

class Main {
    static int[] bam;
    static int[] sadari;
    static int[] dist;
    static int[] dx = {1, 2, 3, 4, 5, 6};
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int n = Integer.parseInt(st.nextToken());
        int m = Integer.parseInt(st.nextToken());

        bam = new int[101];
        sadari = new int[101];
        dist = new int[101];

        Arrays.fill(dist, -1);
        
        for(int i=0; i<n; i++) {
            st = new StringTokenizer(br.readLine());
            int x = Integer.parseInt(st.nextToken());
            int y = Integer.parseInt(st.nextToken());
            sadari[x] = y;
        }

        for(int i=0; i<m; i++) {
            st = new StringTokenizer(br.readLine());
            int u = Integer.parseInt(st.nextToken());
            int v = Integer.parseInt(st.nextToken());
            bam[u] = v;
        }

        bfs(1);
        System.out.print(dist[100]);
    }

    static void bfs(int start) {
        Queue<Integer> q = new LinkedList<>();
        q.add(start);
        dist[start] = 0;
        while(!q.isEmpty()) {
            int cur = q.poll();
            for(int dir=0; dir<6; dir++) {
                int nx = cur + dx[dir];
                if(nx > 100) continue;
                if(sadari[nx] != 0) {
                    nx = sadari[nx];
                } else if(bam[nx] != 0) {
                    nx = bam[nx];
                }

                if(dist[nx] == -1) {
                    q.add(nx);
                    dist[nx] = dist[cur] + 1;
                }
                //if(nx == 100) {
                //    q.add(nx);
                //    dist[nx] = dist[cur] + 1;
                //}
            }
        }
    }
}